File: javascript/src/agents/judge/string-deduplicator.ts
/**
 * Tracks seen strings and replaces duplicates with markers.
 * Only operates on strings; does not handle traversal.
 */
export class StringDeduplicator {
  private readonly seen = new Map<string, boolean>();
  private readonly threshold: number;

  constructor(params: { threshold: number }) {
    this.threshold = params.threshold;
  }

  /**
   * Resets seen strings for a new digest.
   */
  reset(): void {
    this.seen.clear();
  }

  /**
   * Processes a string, returning duplicate marker if seen before.
   * @param str - String to process
   * @returns Original string or duplicate marker
   */
  process(str: string): string {
    if (str.length < this.threshold) return str;

    const key = this.normalize(str);
    if (this.seen.has(key)) return "[DUPLICATE - SEE ABOVE]";

    this.seen.set(key, true);
    return str;
  }

  /**
   * Normalizes string for comparison (whitespace, case).
   */
  private normalize(str: string): string {
    return str
      .replace(/\\[nrt]/g, " ") // JSON-escaped whitespace
      .replace(/[\n\r\t]/g, " ") // Actual whitespace
      .replace(/\s+/g, " ") // Collapse
      .trim()
      .toLowerCase();
  }
}

File: python/scenario/_judge/string_deduplicator.py
"""
String deduplication for reducing token usage in digests.
"""

import re
from typing import Dict


class StringDeduplicator:
    """
    Tracks seen strings and replaces duplicates with markers.

    Only operates on strings; does not handle traversal.
    """

    def __init__(self, *, threshold: int) -> None:
        """
        Args:
            threshold: Minimum string length to consider for deduplication
        """
        self._seen: Dict[str, bool] = {}
        self._threshold = threshold

    def reset(self) -> None:
        """Resets seen strings for a new digest."""
        self._seen.clear()

    def process(self, s: str) -> str:
        """
        Processes a string, returning duplicate marker if seen before.

        Args:
            s: String to process

        Returns:
            Original string or duplicate marker
        """
        if len(s) < self._threshold:
            return s

        key = self._normalize(s)
        if key in self._seen:
            return "[DUPLICATE - SEE ABOVE]"

        self._seen[key] = True
        return s

    def _normalize(self, s: str) -> str:
        """Normalizes string for comparison (whitespace, case)."""
        # JSON-escaped whitespace
        result = re.sub(r"\\[nrt]", " ", s)
        # Actual whitespace
        result = re.sub(r"[\n\r\t]", " ", result)
        # Collapse
        result = re.sub(r"\s+", " ", result)
        return result.strip().lower()
